fsspec/config.py

```diff
@@ -21,12 +26,24 @@ def set_conf_env(conf_dict, envdict=os.environ):
     envdict : dict-like(str, str)
         Source for the values - usually the real environment
     """
+    kwarg_keys = []
     for key in envdict:
-        if key.startswith("FSSPEC"):
-            if key.count("_") < 2:
+        if key.startswith("FSSPEC_") and len(key) > 7 and key[7] != "_":
+            if key.count("_") > 1:
+                kwarg_keys.append(key)
                 continue
-            _, proto, kwarg = key.split("_", 2)
-            conf_dict.setdefault(proto.lower(), {})[kwarg.lower()] = envdict[key]
+            try:
+                value = json.loads(envdict[key])
+            except json.decoder.JSONDecodeError:
+                pass
+            else:
+                if isinstance(value, dict):
+                    _, proto = key.split("_", 1)
+                    conf_dict.setdefault(proto.lower(), {}).update(value)
```

Comment:
If your assumption about the env vars always being a dict at the toplevel is violated, it would also be good to log rather than silently ignore.

Reply:
Also now a warning.